### ISSUE TITLE
channel torrents stuck at 98% downloading

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#5295]
multiple channels are stuck downloading. One with 69 stuck downloaders another with 41 stuck downloaders. Sadly this is blocking a key part of the user experience. 
Nothing can be done by the user to get "unstuck". Due to the inducement of frustration I'm afraid this is something we have to fix before we declare it stable.

minor:
4 minute of a black loading screen on first startup. if you used Tribler before it should warn you: converting data, please wait a few minutes.

**Comment by ichorid** [comment]
This happens because someone created a channel and corrupted their after someone else started downloading it but before they were able to finish. It is now up to the channel's owner (or their Tribler instance) to notice this and fix the situation.

Currently, Tribler performs does basic checks at startup for channel consistency between torrent form and database form. The bad news is it does not check for garbled files in the personal channel torrent. The good news is it is trivial to add. Then again, the bad news is re-committing the channel to a torrent is not enough, as it should be completely regenerated. The good news is that there is a tested routine for consolidating and regenerating a channel.

The bottom line is: if a user produces a corrupted channel torrent, other users can't do anything about it, except possibly not subscribing to that channel.

I will create a PR for stricter checking of personal channel torrents.

**Comment by synctext** [comment]
If channels owners upgrade to 7.5 we need to make sure this problem goes away. Automated channels repair would fix everything right? What do we do with the 100+ most loyal users who are stuck on a channel?

> This happens because someone created a channel and corrupted their after someone else started downloading it but before they were able to finish.

Multiple channels seem to have this issue. Can you please elaborate on what corruption you are suspecting? The PonyORM system? The data structures you created? How can this happen for multiple channels?

Please try to re-create the exact reason why channels are getting corrupted. Or investigate a stuck channel in detail. Then do a targeted fix and repair.

**Comment by ichorid** [comment]
I found the bug. It was there since the beginning.

The bug is only triggered be a very specific condition:
the user must delete a torrent from their channel, and then the deletion command must be ordered to be serialized the last in the corresponding `.mdblob` chunk. In that case, the blob's filename will be the same as the deleted torrent's timestamp. If there already exists an `.mdblob` with the same name, it will be overwritten. This will corrupt the channel torrent.

Now, to figure out how to fix this with minimal changes...

**Comment by synctext** [comment]
wow. Will that overwrite also result in missing pieces from seeding?

**Comment by ichorid** [comment]
Yes. I am working on a unit test to catch that.

**Comment by ichorid** [comment]
Closing this. The fix should do its work on the owner's side, though.
